Jds mempool enhancement 3

lorbax opened the pull request:
apply part of the suggestions to this PR
[https://github.com/stratum-mining/stratum/pull/772](https://github.com/stratum-mining/stratum/pull/772)

lorbax committed:
Move task on main

As suggested in
https://github.com/stratum-mining/stratum/pull/772#issuecomment-1999112198
A task that before was blocking now is moved as a separate task on main

lorbax committed:
Verify that all txs in job are present

As suggested in
https://github.com/stratum-mining/stratum/pull/772#issuecomment-1999112198
This commit introduces a verification that all the transactions in a job
are correctly recognized when a submit block message appears.

lorbax committed:
Error management

Error management when some transactions are missing. The jds should not
break. Instead, tries to recover it by triggering the jds mempool to
retrieve the transactions from its bitcoin node

Fi3 commented on roles/jd-server/src/lib/mempool/mod.rs:
```diff
@@ -58,14 +59,34 @@ impl JDsMempool {
+        // retrieved from the jd client
+        for txid in txids {
+            let transaction = client
+                .get_raw_transaction(&txid.to_string(), None)
```
before fetching txs you should check if they are already in the mempool

Fi3 replied:
another option could  be to use an HashSet right after the receiver

lorbax replied:
I am struggling to figure out a useful case for HashSet. If you propose to use it instead HashMap, then you cant store transactions only by their IDs. What do you mean with "right after the receiver"? Can you be a bit more specific? ^^

BTW already fixed in [this commit](https://github.com/lorbax/stratum/commit/d99d438268660bf31bc4d10cd7c30f7244887141)

lorbax committed:
Fix bug in SendTo output from message_handler

In the message handler the message the triggers the JDS mempool to fill
the transactions is implemented in a bad way. Now it is fixed

Add some documentation

lorbax committed:
Check if a tx is already present in mempool

apply Fi3 suggestion
https://github.com/GitGab19/stratum/pull/16#discussion_r1532513707

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
+        // the trasnactions sent to the mempool can be freed
+        let _ = self_mutex.safe_lock(|a| {
+            a.add_txs_to_mempool.add_txs_to_mempool_inner = AddTrasactionsToMempoolInner {
+                known_transactions: vec![],
```
better to use clear, no need to deallocate if we are going to fill it again

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -231,9 +222,54 @@ impl JobDeclaratorDownstream {
+                                    }
+                                    JobDeclaration::SubmitSolution(_) => todo!(),
+                                }
+                                Self::send(self_mutex.clone(), m).await.unwrap();
```
this always send the message also for the error cases

Fi3 replied:
if we get one of the error cases above something really strange is happening better to close the connection. Downstream will decide what to do

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
+        Ok(transactions_list)
+    }
+
+    async fn send_txs_to_mempool(self_mutex: Arc<Mutex<Self>>) {
```
I think that you can use an &Arc<Mutex<Seld>> here

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
+        Ok(hex::encode(serialize(&block)))
+    }
+
+    fn are_all_job_transactions_present(
```
change the function name i something else, this seems just a condition check but it do collect txs.

you can use colletc_txs_for_job

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
+            .unwrap();
+        let sender_add_txs_to_mempool = add_txs_to_mempool.sender_add_txs_to_mempool;
+        let add_txs_to_mempool_inner = add_txs_to_mempool.add_txs_to_mempool_inner;
+        let _ = sender_add_txs_to_mempool
```
why you ignore the error here?

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
+            .send(add_txs_to_mempool_inner)
+            .await;
+        // the trasnactions sent to the mempool can be freed
+        let _ = self_mutex.safe_lock(|a| {
```
same

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -247,55 +283,102 @@ impl JobDeclaratorDownstream {
+                                                    .clone()
+                                                    .safe_lock(|a| a.mempool.clone())
+                                                    .unwrap();
+                                                tokio::select! {
```
seems that you want to use a tokio::timeout here

Fi3 commented on roles/jd-server/src/lib/job_declarator/mod.rs:
```diff
@@ -408,6 +491,8 @@ impl JobDeclarator {
                     sender.clone(),
                     &config,
                     mempool.clone(),
+                    // each downstream has its own sender (multi producer single consumer)
```
remove this comment you see it from the type
